### PR TITLE
Shimming legacy .NET worker JsonFunctionProvider to ensure backwards compatibility.

### LINKED ISSUE #10410: Dependency injection fails for oop-host when using a .NET Isolated payload with the `worker.sdk` package version 1.3.0 or earlier.
When using a .NET Isolated payload with [Microsoft.Azure.Functions.Worker.sdk](https://www.nuget.org/packages/Microsoft.Azure.Functions.Worker.sdk) version 1.3.0 or below, host will fail with the following exception.

> System.InvalidOperationException: 'Unable to resolve service for type 'System.String' while attempting to activate 'Microsoft.Azure.WebJobs.Extensions.FunctionMetadataLoader.JsonFunctionProvider'.'

Stack trace

>    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.CreateArgumentCallSites(ServiceIdentifier serviceIdentifier, Type implementationType, CallSiteChain callSiteChain, ParameterInfo[] parameters, Boolean throwIfCallSiteNotFound)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.CreateConstructorCallSite(ResultCache lifetime, ServiceIdentifier serviceIdentifier, Type implementationType, CallSiteChain callSiteChain)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.TryCreateExact(ServiceDescriptor descriptor, ServiceIdentifier serviceIdentifier, CallSiteChain callSiteChain, Int32 slot)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.TryCreateEnumerable(ServiceIdentifier serviceIdentifier, CallSiteChain callSiteChain)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.CreateCallSite(ServiceIdentifier serviceIdentifier, CallSiteChain callSiteChain)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.GetCallSite(ServiceIdentifier serviceIdentifier, CallSiteChain callSiteChain)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateServiceAccessor(ServiceIdentifier serviceIdentifier)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(ServiceIdentifier serviceIdentifier, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType)
   at Microsoft.Azure.WebJobs.Script.WebHost.WebJobsScriptHostService.GetService(Type serviceType) in D:\src\azure-functions-host\src\WebJobs.Script.WebHost\WebJobsScriptHostService.cs:line 962

- The issue does not happen for version 1.5.0-preview1 or above.
- Issue happens only on dev branch which uses .NET DI framework.
- No issues with in-proc branch which uses DryIOC.

JsonFunctionProvider constructor was changed between the working and non-working version.

1.30.0

image

1.5.0-preview1

image

We believe DryIOC handles this by passing null/empty string value while .NET DI framework fails. We need to fix the oop host branch to handle this service resolution.

### PR DESCRIPTION
**Pull request by fabiocav**
### Issue describing the changes in this PR

resolves #10410

### Pull request checklist

**IMPORTANT**: Currently, changes must be backported to the `in-proc` branch to be included in Core Tools and non-Flex deployments.

* [x] Backporting to the `in-proc` branch is not required
    * Otherwise: Link to backporting PR
* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [x] I have added all required tests (Unit tests, E2E tests)

### CODE REVIEW

File: src/WebJobs.Script.WebHost/DependencyInjection/JobHostScopedServiceProviderFactory.cs [hunk 1]
```diff
@@ -76,12 +76,69 @@ public IServiceProvider CreateServiceProvider(IServiceCollection services)
...
 
+        private void ShimBreakingChanges(IServiceCollection services, ILogger logger)
+        {
+            ShimActivatorUtilitiesConstructorAttributeTypes(services, logger);
+            ShimLegacyNetWorkerMetadataProvider(services, logger);
+        }
+
+        /// <summary>
+        /// Versions 1.3.0 and older of Microsoft.Azure.Functions.Worker.Sdk register a type with a constructor that
+        /// takes a strign parameter. This workeed with DryIoc, but not with the new DI container.
```
Comment (kshyju) [inline src/WebJobs.Script.WebHost/DependencyInjection/JobHostScopedServiceProviderFactory.cs]: nit: typos  (string, worked)

File: src/WebJobs.Script.WebHost/DependencyInjection/JobHostScopedServiceProviderFactory.cs [hunk 2]
```diff
@@ -76,12 +76,69 @@ public IServiceProvider CreateServiceProvider(IServiceCollection services)
...
+        /// Shimming those types to provide backwards compatibility, but this should be removed in the future.
+        /// </summary>
+        /// <param name="services">The <see cref="ServiceCollection"/> to inspect and modify.</param>
+        private static void ShimLegacyNetWorkerMetadataProvider(IServiceCollection services, ILogger logger)
```
Comment (kshyju) [inline src/WebJobs.Script.WebHost/DependencyInjection/JobHostScopedServiceProviderFactory.cs:92]: nit:  missing `logger` in method summary.

**Review by kshyju — APPROVED** [review]
LGTM.

File: src/WebJobs.Script.WebHost/DependencyInjection/JobHostScopedServiceProviderFactory.cs (hunk 2 above)
Comment (fabiocav) [reply src/WebJobs.Script.WebHost/DependencyInjection/JobHostScopedServiceProviderFactory.cs:92]: Good catch. This was a late addition. Also updated `ShimActivatorUtilitiesConstructorAttributeTypes` as I added the parameter there as well.

**Review by brettsam — APPROVED** [review]
I know this is coming in hot -- but can we make sure to file an issue for a follow-up test?

**Comment by fabiocav** [comment]
Created this issue to track: https://github.com/Azure/azure-functions-host/issues/10414

**Pull request merged**